README.md

```diff
@@ -4,222 +4,1008 @@
+        errorIfFalse = "Expected file to have extention of "~this.wantedExtention;
+
+        return arg.endsWith(this.wantedExtention);
+    }
+}
+
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    @HasExtention(".json")
+    File file;
+
+    // omitted...
+}
+```
+
+To start, we create a struct called `HasExtention` and we give it a field member called `string wantedExtention;`.
+
+Before I continue, I want to explicitly state that this validator wants to perform validation on the raw string that the user provides (pre-arg-binded) and *not* on the final value (post-arg-binded). This is referred to as "Pre Validation". So on that note...
+
+Next, and most importantly, we define a function that specifically called `onPreValidate` that follows the following convention:
+
+```d
+bool onPreValidate(string arg, ref string error);
+```
+
+This is the function that performs the actual validation (in this case, "Pre" validation).
+
+It returns `true` if there are no validation errors, otherwise it returns `false` and optionally sets the `error` string to a user-friendly error (one is automatically generated otherwise).
+
+The first parameter is the raw string that the user has provided us, and I just explained the second parameter.
+
+So for our `HasExtention` validator, all we do is check if the user's file path ends with `this.wantedExtention`, which we set the value of later.
+
+Now, inside `CatCommand` all we've done is attach our `HasExtention` struct as a UDA (and if you're not familiar with D, congrats, you just made your first UDA!). JCLI will automatically detect that `@HasExtention` is a validator because it follows that convention mentioned just above.
+
+Because D is wonderful, it will automatically generate a constructor for us where the first parameter sets the `wantedExtention` member. So `@HasExtention(".json")` will set the extention we want to `".json"`.
+
+And that's literally all there is to it, let's test:
+
+```bash
+# Passing
+$> ./mytool cat ./dub.json
+[contents of dub.json since validation was a success]
+Program exited with status code 0
+
+# Failing
+$> ./mytool cat ./.gitignore
+ERROR: For positional arg 0(filePath): Expected file to have extention of .json
+Program exited with status code -1
+```
+
+The other type of validation is post-arg-binded validation, which performs validation on the final value provided by an arg binder.
+
+Let's make a validator that ensures that the file is under a certain size:
+
+```d
+struct MaxSize
+{
+    ulong maxSize;
+
+    bool onValidate(File file, ref string error)
+    {
+        error = "File is too large.";
+        return file.size() <= this.maxSize;
+    }
+}
+
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    @HasExtention(".json")
+    @MaxSize(2)
```

Comment:
While it does seem to be a benefit for the `@PreValidate` and `@PostValidate` style, I'm pretty certain this can be replicated with the current system already, without losing too much useability.

```d
// Untested, but should give you the idea.
//
// Writing in a style that assumes something like issue #12 is already implemented.
struct PostValidate(Type, alias Func)
{
    static assert(is(Func == ValidationResult delegate(Type value)), "Some user friendly message here.");

    ValidationResult onValidate(Type value)
    {
        return Func(value);
    }
}
```

Reply:
Used as:

```d
@PostValidate!File(file => file.size() <= 2)
```

Reply:
Or maybe even:

```d
@PostValidate!((File file) => file.size() <= 2)
```

Since you'd be able to introspect the parameter's type

Reply:
May as well. #24